crates/chain/src/keychain/txout_index.rs

```diff
@@ -330,98 +475,115 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+                .all_spks()
+                // This range is keeping only the spks with descriptor_id equal to
+                // `descriptor_id`. We don't use filter here as range is more optimized.
+                .range((descriptor_id, u32::MIN)..(descriptor_id, u32::MAX))
+                .last()
+                .map_or(0, |((_, index), _)| *index + 1),
+        )
     }
 
-    /// Get an unbounded spk iterator over a given `keychain`.
-    ///
-    /// # Panics
-    ///
-    /// This will panic if the given `keychain`'s descriptor does not exist.
-    pub fn unbounded_spk_iter(&self, keychain: &K) -> SpkIterator<Descriptor<DescriptorPublicKey>> {
-        SpkIterator::new(
-            self.keychains
-                .get(keychain)
-                .expect("keychain does not exist")
-                .clone(),
-        )
+    /// Get an unbounded spk iterator over a given `keychain`. Returns `None` if the provided
+    /// keychain doesn't exist
+    pub fn unbounded_spk_iter(
+        &self,
+        keychain: &K,
+    ) -> Option<SpkIterator<Descriptor<DescriptorPublicKey>>> {
+        let descriptor = self.descriptor_of_keychain(keychain)?.1.clone();
+        Some(SpkIterator::new(descriptor))
     }
 
     /// Get unbounded spk iterators for all keychains.
     pub fn all_unbounded_spk_iters(
         &self,
     ) -> BTreeMap<K, SpkIterator<Descriptor<DescriptorPublicKey>>> {
-        self.keychains
+        self.keychains_to_descriptors
             .iter()
-            .map(|(k, descriptor)| (k.clone(), SpkIterator::new(descriptor.clone())))
+            .map(|(k, (_, descriptor))| (k.clone(), SpkIterator::new(descriptor.clone())))
             .collect()
     }
 
     /// Iterate over revealed spks of all keychains.
     pub fn revealed_spks(&self) -> impl DoubleEndedIterator<Item = (K, u32, &Script)> + Clone {
-        self.keychains.keys().flat_map(|keychain| {
-            self.revealed_keychain_spks(keychain)
-                .map(|(i, spk)| (keychain.clone(), i, spk))
-        })
+        self.keychains_to_descriptors
+            .keys()
+            .filter_map(|keychain| {
+                self.revealed_keychain_spks(keychain)
+                    .map(|spk_iter| spk_iter.map(|(i, spk)| (keychain.clone(), i, spk)))
+            })
+            .flatten()
     }
 
-    /// Iterate over revealed spks of the given `keychain`.
+    /// Iterate over revealed spks of the given `keychain`. Returns None if the provided `keychain`
+    /// doesn't exist.
     pub fn revealed_keychain_spks(
         &self,
         keychain: &K,
-    ) -> impl DoubleEndedIterator<Item = (u32, &Script)> + Clone {
-        let next_i = self.last_revealed.get(keychain).map_or(0, |&i| i + 1);
-        self.inner
-            .all_spks()
-            .range((keychain.clone(), u32::MIN)..(keychain.clone(), next_i))
-            .map(|((_, i), spk)| (*i, spk.as_script()))
+    ) -> Option<impl DoubleEndedIterator<Item = (u32, &Script)> + Clone> {
```

Comment:
:question: What's your opinion on https://github.com/bitcoindevkit/bdk/pull/1203#discussion_r1495082281 ? I still think that this should just be an empty iterator -- it's more ergonomic and technically the right answer.

Reply:
I am very torn here.
On one hand - of course, returning an Option is not ergonomic at all, and that's mostly the reason why I decided to panic at first. And I agree that returning an empty iterator is technically the right answer, but at the same time I wonder how many users will want to know that they've asked for a keychain we don't have vs how many don't care (and if you don't care but we tell you, it's easy to just discard the option, even though your code becomes not_so_clean, vs if you do care and we don't tell you, that's a problem).

An example comes to my mind: our `Wallet::sign` doesn't return any Error if you don't have a private key, it just returns back the transaction unsigned. I'm not advocating for changing this - but we had many users just telling us "there's a bug! I call `sign` and nothing happens!" because our technically right answer didn't communicate what was going on underneath ("we tried to sign your transaction but we simply don't have a private key").

Now, it doesn't mean that we need to solve this in code - maybe decent docs are enough for this?

I don't like the Options either, and that's mostly why I pushed a separate commit for the panic -> option change, so that I could simply remove it if we decide that's not the right way to go. I can remove the option from the methods that return an iterator, leave it everywhere else, and document clearly that if your keychain isn't inserted the returned iterator is going to be empty. If we realize that this confuses our users, we can always re-introduce the option in a later version.

Reply:
Hey @danielabrozzoni  I think you answered your own questions here, which is to return an empty iterator, the technically right way, but clearly document the behavior. For these internal chain crate APIs I think it's ok to lean in favor of technically right and avoiding panics. Anyone using the chain crate should be more sophisticated than our `wallet` API users who would need more explicit return values.

Reply:
I'm leaning technically left.

Reply:
I really need to decide what to do here before rebasing, since I need to rebase on top of https://github.com/bitcoindevkit/bdk/commit/fac228337c79e1a7c406500c177e049e7f9f5da7 and I don't want to redo the rebase.

Returning an empty iterator is the easiest solution to implement, so I'll go with that for now, and open an issue to track the fact that some of us liked the option better. We can always switch to option in a subsequent pr.